pkg/wal/wal.go

```diff
@@ -42,19 +72,642 @@ type Segment interface {
+
+	for seriesID, timestamps := range buffer.timestampMap {
+		// Generate seriesID binary
+		seriesIDBytes := seriesID.Marshal()
+		seriesIDBytesLen := len(seriesIDBytes)
+
+		// Generate timestamps compression binary
+		log.timestampsBuffer.Reset()
+		timestampWriter := encoding.NewWriter()
+		timestampEncoder := encoding.NewXOREncoder(timestampWriter)
+		timestampWriter.Reset(log.timestampsBuffer)
+		for _, timestamp := range timestamps {
+			timestampEncoder.Write(timeTouUnixNano(timestamp))
+		}
+		timestampWriter.Flush()
+		timestampsBytes := log.timestampsBuffer.Bytes()
+		timestampsBytesLen := len(timestampsBytes)
+
+		// Generate values compression binary
+		valuesBytes := snappy.Encode(nil, buffer.valueMap[seriesID])
+
+		// Write entry data
+		entryLength := seriesIDLength + seriesIDBytesLen + seriesCountLength + timestampsBinaryLength + timestampsBytesLen + len(valuesBytes)
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, int64(entryLength))
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, int16(seriesIDBytesLen))
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, seriesIDBytes)
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, int32(len(timestamps)))
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, int16(timestampsBytesLen))
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, timestampsBytes)
+		log.writeToBytesBuffer(log.bytesBuffer, binary.LittleEndian, valuesBytes)
+	}
+	// Rewrite batch length
+	batchBytes := log.bytesBuffer.Bytes()
+	batchLength = len(batchBytes) - batchWriteLength
+	rewriteInt64InBuf(batchBytes, int64(batchLength), 0, binary.LittleEndian)
+	log.bytesBuffer.Reset()
+
+	// Flush
+	_, err := log.workSegment.file.Write(batchBytes)
+	if err != nil {
+		log.logger.Error().Err(err).Msg("Write WAL segment file error, file: " + log.workSegment.path)
+		buffer.notifyRequests(err)
+		return
+	}
+	err = log.workSegment.file.Sync()
+	if err != nil {
+		log.logger.Error().Err(err).Msg("Sync WAL segment file to disk error, file: " + log.workSegment.path)
+		buffer.notifyRequests(err)
+		return
+	}
+	buffer.notifyRequests(nil)
+	if log.logger.Debug().Enabled() {
+		log.logger.Debug().Msg("Flushed buffer to WAL. file: " + log.workSegment.path +
+			", elements: " + strconv.Itoa(buffer.count) +
+			", bytes: " + strconv.Itoa(len(batchBytes)))
+	}
+}
+
+func (log *Log) newBuffer() {
+	log.buffer.timestampMap = make(map[common.SeriesIDV2][]time.Time)
+	log.buffer.valueMap = make(map[common.SeriesIDV2][]byte)
+	log.buffer.callbackMap = make(map[common.SeriesIDV2][]func(common.SeriesIDV2, time.Time, []byte, error))
+	log.buffer.count = 0
+}
+
+func (log *Log) load() error {
+	files, err := os.ReadDir(log.path)
+	if err != nil {
+		return errors.Wrap(err, "Can not read dir: "+log.path)
+	}
+	// Load all of WAL segments.
+	var workSegmentID SegmentID
+	log.segmentIndexMap = make(map[SegmentID]*segment)
+	for _, file := range files {
```

Comment:
During the booting process, it is necessary to remove all files except for the `workSegment`. 

To create a new `workSegment`, the user should rotate the WAL and delete the previous one. In case the deletion fails due to a process crash, the `load` function will take care of cleaning up all non-work segments.